src/libraries/System.Private.CoreLib/src/System/Int128.cs

```diff
@@ -0,0 +1,1832 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    /// <summary>Represents a 128-bit signed integer.</summary>
+    [Intrinsic]
+    [StructLayout(LayoutKind.Sequential)]
+    public readonly struct Int128
+        : IBinaryInteger<Int128>,
+          IMinMaxValue<Int128>,
+          ISignedNumber<Int128>
+    {
+        internal const int Size = 16;
+
+        // Unix System V ABI actually requires this to be little endian
+        // order and not `upper, lower` on big endian systems.
+
+        private readonly ulong _lower;
+        private readonly ulong _upper;
+
+        /// <summary>Initializes a new instance of the <see cref="Int128" /> struct.</summary>
+        /// <param name="upper">The upper 64-bits of the 128-bit value.</param>
+        /// <param name="lower">The lower 64-bits of the 128-bit value.</param>
+        [CLSCompliant(false)]
+        public Int128(ulong upper, ulong lower)
+        {
+            _lower = lower;
+            _upper = upper;
+        }
+
+        internal ulong Lower => _lower;
+
+        internal ulong Upper => _upper;
+
+        /// <inheritdoc cref="IComparable.CompareTo(object)" />
+        public int CompareTo(object? value)
+        {
+            if (value is Int128 other)
+            {
+                return CompareTo(other);
+            }
+            else if (value is null)
+            {
+                return 1;
+            }
+            else
+            {
+                throw new ArgumentException(SR.Arg_MustBeInt128);
+            }
+        }
+
+        /// <inheritdoc cref="IComparable{T}.CompareTo(T)" />
+        public int CompareTo(Int128 value)
+        {
+            if (this < value)
+            {
+                return -1;
+            }
+            else if (this > value)
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+        }
+
+        /// <inheritdoc cref="object.Equals(object?)" />
+        public override bool Equals([NotNullWhen(true)] object? obj)
```

Comment:
Wouldn't a check with `GetType()` make more sense here?
See https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aaGfef6m5JA2AQIAGwYBRAI4BXbONwBBABQRgAKxhgMDDZoCU3Pr07VTlpgHYG6rQwCWuBuIgA7AOb6MACxhQhgDcJpYAvqGmAsRCImKSsgpKAEL22rr6WsYWVuZWUbYGAHQA4jAYACoAngAOMKqGDAC8TQwYtTAQAGaqbl7BkXwR1GFAA===

Reply:
No. The `obj` must be unboxed if the type is `Int128` and a comparison must then be done using the unboxed value.

`x is T other` is the standard pattern for that and the runtime should optimize it accordingly.